l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,638 @@
+    }
+
+    /// @dev Transfers tokens from the depositor address to the smart contract address.
+    /// @return The difference between the contract balance before and after the transferring of funds.
+    function _depositFunds(address _from, IERC20 _token, uint256 _amount) internal returns (uint256) {
+        uint256 balanceBefore = _token.balanceOf(address(this));
+        _token.safeTransferFrom(_from, address(this), _amount);
+        uint256 balanceAfter = _token.balanceOf(address(this));
+
+        return balanceAfter - balanceBefore;
+    }
+
+    /// @notice Initiates a deposit transaction within Bridgehub, used by `requestL2TransactionTwoBridges`.
+    function bridgehubDeposit(
+        uint256 _chainId,
+        address _prevMsgSender,
+        uint256, // l2Value, needed for Weth deposits in the future
+        bytes calldata _data
+    ) external payable override onlyBridgehub returns (L2TransactionRequestTwoBridgesInner memory request) {
+        require(l2BridgeAddress[_chainId] != address(0), "ShB l2 bridge not deployed");
+
+        (address _l1Token, uint256 _depositAmount, address _l2Receiver) = abi.decode(
+            _data,
+            (address, uint256, address)
+        );
+        require(_l1Token != l1WethAddress, "ShB: WETH deposit not supported");
+        require(bridgehub.baseToken(_chainId) != _l1Token, "ShB: baseToken deposit not supported");
+
+        uint256 amount;
+        if (_l1Token == ETH_TOKEN_ADDRESS) {
+            amount = msg.value;
+            require(_depositAmount == 0, "ShB wrong withdraw amount");
+        } else {
+            require(msg.value == 0, "ShB m.v > 0 for BH d.it 2");
+            amount = _depositAmount;
+
+            uint256 withdrawAmount = _depositFunds(_prevMsgSender, IERC20(_l1Token), _depositAmount);
+            require(withdrawAmount == _depositAmount, "5T"); // The token has non-standard transfer logic
+        }
+        require(amount != 0, "6T"); // empty deposit amount
+
+        bytes32 txDataHash = keccak256(abi.encode(_prevMsgSender, _l1Token, amount));
+        if (!hyperbridgingEnabled[_chainId]) {
+            chainBalance[_chainId][_l1Token] += amount;
+        }
+
+        {
+            // Request the finalization of the deposit on the L2 side
+            bytes memory l2TxCalldata = _getDepositL2Calldata(_prevMsgSender, _l2Receiver, _l1Token, amount);
+
+            request = L2TransactionRequestTwoBridgesInner({
+                magicValue: TWO_BRIDGES_MAGIC_VALUE,
+                l2Contract: l2BridgeAddress[_chainId],
+                l2Calldata: l2TxCalldata,
+                factoryDeps: new bytes[](0),
+                txDataHash: txDataHash
+            });
+        }
+        emit BridgehubDepositInitiatedSharedBridge(_chainId, txDataHash, _prevMsgSender, _l2Receiver, _l1Token, amount);
+        // kl todo. We are breaking the previous events here, as we don't have the txHash, so we can't emit a DepositInitiated event
```

Comment:
Please don't leave stray todo comments in code (especially one that is up for an audit pretty soon)